File: 26_Dynamic Programming-II/M-TUF-Super Egg Drop.java
// Super Egg Drop
// https://takeuforward.org/plus/dsa/problems/super-egg-drop

/**
 * You are given n identical eggs and a building with k floors numbered from 1 to k.
 * There exists an unknown floor f such that:
 * Any egg dropped from a floor higher than f will break.
 * Any egg dropped from f or below will not break.
 * Your goal is to determine the value of f using the minimum number of egg drops in the worst-case scenario.
 *
 * You may assume:
 * An egg that survives a fall can be reused.
 * A broken egg must be discarded.
 * All eggs behave the same way.
 * If an egg breaks at a floor, it would break at any higher floor.
 * If an egg survives a floor, it would survive any lower floor.
 *
 * Examples:
 * Input: n = 2, k = 36
 * Output: 8
 * Explanation: Minimum 8 moves are needed in the worst-case using binary + linear strategy.
 */

class Solution {
    public int eggDrop(int eggs, int floors) {
        // Create a DP table where dp[i][j] represents the minimum number of attempts
        // needed for i eggs and j floors
        int[][] dp = new int[eggs + 1][floors + 1];
        
        // Base cases:
        // 1. If there's only 1 egg, we need to check each floor one by one
        for (int j = 1; j <= floors; j++) {
            dp[1][j] = j;
        }
        
        // 2. If there's only 1 floor, we need only 1 attempt
        // 3. If there are 0 floors, we need 0 attempts
        for (int i = 1; i <= eggs; i++) {
            dp[i][1] = 1;
            dp[i][0] = 0;
        }
        
        // Fill the DP table
        for (int i = 2; i <= eggs; i++) {          // For each number of eggs
            for (int j = 2; j <= floors; j++) {     // For each number of floors
                dp[i][j] = Integer.MAX_VALUE;
                
                // Try dropping egg from each floor k from 1 to j
                for (int k = 1; k <= j; k++) {
                    // If egg breaks, check floors below (i-1 eggs, k-1 floors)
                    // If egg doesn't break, check floors above (i eggs, j-k floors)
                    int attempts = 1 + Math.max(dp[i-1][k-1], dp[i][j-k]);
                    if (attempts < dp[i][j]) {
                        dp[i][j] = attempts;
                    }
                }
            }
        }
        
        return dp[eggs][floors];
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        int n = 2, k = 36;
        System.out.println("Minimum number of attempts needed: " + sol.eggDrop(n, k));
    }
}

File: JavaScript/26_Dynamic Programming-II/M-TUF-Super Egg Drop.js
/**
 * TUF: Super Egg Drop (Egg Dropping Problem)
 * https://leetcode.com/problems/super-egg-drop/
 *
 * Find minimum number of moves to determine the critical floor.
 *
 * Type             | Details
 * -----------------+----------------------------------------------------
 * Time Complexity  | O(K * N * log N) – Binary search optimization
 * Space Complexity | O(K * N) – DP table
 */

/**
 * Super Egg Drop - O(K * N²) approach
 * @param {number} k - Number of eggs
 * @param {number} n - Number of floors
 * @returns {number}
 */
function superEggDropBasic(k, n) {
    // dp[i][j] = minimum moves with i eggs and j floors
    const dp = Array.from({ length: k + 1 }, () => new Array(n + 1).fill(0));
    
    // Base cases
    for (let j = 1; j <= n; j++) {
        dp[1][j] = j; // With 1 egg, need j moves for j floors
    }
    for (let i = 1; i <= k; i++) {
        dp[i][1] = 1; // With 1 floor, need 1 move
    }
    
    for (let i = 2; i <= k; i++) {
        for (let j = 2; j <= n; j++) {
            dp[i][j] = Infinity;
            for (let x = 1; x <= j; x++) {
                // Drop egg from floor x
                // If breaks: check floors below with i-1 eggs
                // If survives: check floors above with i eggs
                const worst = 1 + Math.max(dp[i - 1][x - 1], dp[i][j - x]);
                dp[i][j] = Math.min(dp[i][j], worst);
            }
        }
    }
    
    return dp[k][n];
}

/**
 * Super Egg Drop - O(K * N * log N) with Binary Search
 * @param {number} k - Number of eggs
 * @param {number} n - Number of floors
 * @returns {number}
 */
function superEggDrop(k, n) {
    const memo = new Map();
    
    function dp(eggs, floors) {
        if (floors <= 1) return floors;
        if (eggs === 1) return floors;
        
        const key = `${eggs},${floors}`;
        if (memo.has(key)) return memo.get(key);
        
        let low = 1, high = floors, result = floors;
        
        while (low <= high) {
            const mid = Math.floor((low + high) / 2);
            
            const broken = dp(eggs - 1, mid - 1);   // Egg breaks
            const survived = dp(eggs, floors - mid); // Egg survives
            
            const worst = 1 + Math.max(broken, survived);
            result = Math.min(result, worst);
            
            if (broken > survived) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        
        memo.set(key, result);
        return result;
    }
    
    return dp(k, n);
}

/**
 * O(K * N) approach - dp[m][k] = max floors we can check with m moves and k eggs
 */
function superEggDropOptimal(k, n) {
    // dp[m][k] = max floors we can check with m moves and k eggs
    // dp[m][k] = dp[m-1][k-1] + dp[m-1][k] + 1
    
    const dp = Array.from({ length: n + 1 }, () => new Array(k + 1).fill(0));
    
    let m = 0;
    while (dp[m][k] < n) {
        m++;
        for (let i = 1; i <= k; i++) {
            dp[m][i] = dp[m - 1][i - 1] + dp[m - 1][i] + 1;
        }
    }
    
    return m;
}

// Test cases
console.log("superEggDrop(1, 2):", superEggDrop(1, 2)); // 2
console.log("superEggDrop(2, 6):", superEggDrop(2, 6)); // 3
console.log("superEggDrop(3, 14):", superEggDrop(3, 14)); // 4

console.log("\nOptimal approach:");
console.log("superEggDropOptimal(2, 6):", superEggDropOptimal(2, 6)); // 3
console.log("superEggDropOptimal(2, 100):", superEggDropOptimal(2, 100)); // 14